### PR TITLE
ROX-19696: Process deduper state

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

When reconciling on sensor side, sensor's gRPC deduper should be populated from central resource hashes.

* Follow-up: After #7755 is merged, we need to check if central has the capabilities before going down the path of setting the deduper state.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

### Here I tell how I validated my change

* [x] Unit tests added.
* [x] CI

Manual tests cannot be done before #7755

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: sensor/common/deduper/deduper.go [hunk 1]
```diff
@@ -1,43 +1,111 @@
...
+		&central.SensorEvent_ComplianceOperatorScanSettingBinding{},
+		&central.SensorEvent_ComplianceOperatorScan{},
+	}
 )
 
-// key is the key by which messages are deduped.
-type key struct {
+// Key is the key by which messages are deduped.
+type Key struct {
 	id           string
```
Comment (fredrb) [inline sensor/common/deduper/deduper.go:46]: Make `id` and `resourceType` publicly accessible as well.

Comment (lvalerom) [reply sensor/common/deduper/deduper.go]: Done

File: sensor/common/sensor/central_communication_impl.go [hunk 2]
```diff
@@ -242,7 +247,41 @@ func (s *centralCommunicationImpl) initialSync(stream central.SensorService_Comm
...
+	}
+
+	if s.clientReconciliation {
+		// If client is reconciling, Sensor needs to wait for Central to send the hashes
+		return s.initialDeduperSync(stream, s.syncTimeout)
+	}
+	return nil
+}
+
+func (s *centralCommunicationImpl) initialDeduperSync(stream central.SensorService_CommunicateClient, timeout time.Duration) error {
```
Comment (fredrb) [inline sensor/common/sensor/central_communication_impl.go:261]: I would suggest to remove this block from this PR in favor or the change here #7755. If you want to test with the actual fetching of deduper state, let's rebase this branch on top of the 7755 one.

File: sensor/common/sensor/central_communication_impl.go [hunk 3]
```diff
@@ -242,7 +247,41 @@ func (s *centralCommunicationImpl) initialSync(stream central.SensorService_Comm
...
+	done := make(chan struct{})
+	var err error
+	var msg *central.MsgToSensor
+	go func() {
+		msg, err = stream.Recv()
+		close(done)
+	}()
+	select {
+	case <-time.After(timeout):
```
Comment (fredrb) [inline sensor/common/sensor/central_communication_impl.go:270]: This is great, do you mind opening it in a separate PR, on top of the changes in #7755? There is also a specific test for it: [ROX-17489](https://issues.redhat.com/browse/ROX-17489).

File: sensor/common/deduper/deduper.go [hunk 4]
```diff
@@ -1,43 +1,128 @@
...
+		&central.SensorEvent_ComplianceOperatorResult{},
+		&central.SensorEvent_ComplianceOperatorProfile{},
+		&central.SensorEvent_ComplianceOperatorRule{},
+		&central.SensorEvent_ComplianceOperatorScanSettingBinding{},
+		&central.SensorEvent_ComplianceOperatorScan{},
+	}
 )
 
+// Key is an interface to the key by which the messages are deduped.
+type Key interface {
```
Comment (fredrb) [inline sensor/common/deduper/deduper.go:45]: Sorry for the miscommunication, but after shuffling a few things in #7981, I think it's actually cleaner to just use an exported `Key` struct type as opposed to an interface. More comments on it below.

File: sensor/common/deduper/deduper.go [hunk 5]
```diff
@@ -1,43 +1,128 @@
...
+		}
+	}
+	return nil, fmt.Errorf("invalid type: %s", typeStr)
+}
+
 // deduper takes care of deduping sensor events.
 type deduper struct {
 	stream   messagestream.SensorMessageStream
-	lastSent map[key]uint64
+	lastSent map[Key]uint64
```
Comment (fredrb) [inline sensor/common/deduper/deduper.go:91]: Interfaces can be `nil` so there is a risk for a panic when doing map operations with a `Key` type. For that reason, a `struct` with primitive values makes more sense to me as a map key.

File: sensor/common/deduper/deduper.go [hunk 6]
```diff
@@ -1,43 +1,128 @@
...
 )
 
+// Key is an interface to the key by which the messages are deduped.
+type Key interface {
+	GetID() string
+	GetType() reflect.Type
+}
+
 // key is the key by which messages are deduped.
 type key struct {
```
Comment (fredrb) [inline sensor/common/deduper/deduper.go:51]: WDYT about moving the all the `key` related functions and structs into their own file? Like `sensor/common/deduper/key.go`?

File: sensor/common/deduper/deduper.go [hunk 7]
```diff
@@ -1,43 +1,128 @@
...
 
+func (k key) GetID() string {
+	return k.id
+}
+
+func (k key) GetType() reflect.Type {
+	return k.resourceType
+}
+
+func keyFrom(v string) (Key, error) {
```
Comment (fredrb) [inline sensor/common/deduper/deduper.go:64]: Could we write some unit tests for this? I'd like to see that the actual examples from deduper states (e.g.: `Binding:d4441ace-ea04-4d5b-8693-97d57f10dace -> 6133788155869413633`) would work.

File: sensor/common/sensor/central_communication_test.go [hunk 8]
```diff
@@ -142,6 +136,105 @@ func (c *centralCommunicationSuite) Test_StopCentralCommunication() {
...
+			for _, msg := range tc.componentMessages {
+				responsesC <- message.New(msg)
+			}
+
+			if tc.eventuallyMatchesState != nil {
+				c.Assert().Eventually(func() bool {
+					mu.Lock()
+					defer mu.Unlock()
+					return tc.eventuallyMatchesState(copySentMessages(sentMessages))
+				}, 3*time.Second, 100*time.Millisecond)
```
Comment (fredrb) [inline sensor/common/sensor/central_communication_test.go:206]: nit: maybe these times are too high for a unit test? Specially now that there are no gRPC connections being actually opened in the host. Maybe a max of `300ms` and a tick of `10ms` would already do the trick? But in reality the threshold might be much lower.

File: sensor/common/sensor/central_communication_impl.go [hunk 9]
```diff
@@ -242,7 +247,41 @@ func (s *centralCommunicationImpl) initialSync(stream central.SensorService_Comm
...
+		log.Error("timeout reached waiting for the initial deduper sync")
+		return errors.Errorf("%v timeout reached waiting for the initial deduper state", timeout)
```
Comment (fredrb) [inline sensor/common/sensor/central_communication_impl.go:272]: The problem with things at this state (without #7755) is that if this case is reached, we get into a fail loop. Sensor will restart the connection here and it will expect for the deduper state (again) and return an error (again).

**Review by fredrb — COMMENTED** [review]
I'm in a conundrum in terms of what to do with this PR. 

1) We could rebase on top of #7755, but them it will be blocked until that is merged.
2) We could remove the duplicate code between this and #7755, but then the tests written here would not pass, since they are testing `central_communication`
3) We could merge this first, but then leave a broken state (e.g. `ROX_SENSOR_RECONCILE=true`) where sensor gets in a fail loop waiting for deduper state. 

I'm leaning towards (1), because it seems like a natural order of things. What would you prefer?

File: sensor/common/sensor/central_communication_impl.go (hunk 2 above)
Comment (lvalerom) [reply sensor/common/sensor/central_communication_impl.go:261]: Done in #8084

File: sensor/common/sensor/central_communication_impl.go (hunk 3 above)
Comment (lvalerom) [reply sensor/common/sensor/central_communication_impl.go:270]: Will do

File: sensor/common/deduper/deduper.go (hunk 5 above)
Comment (lvalerom) [reply sensor/common/deduper/deduper.go:91]: Makes sense. I've changed it in #8078

File: sensor/common/deduper/deduper.go (hunk 6 above)
Comment (lvalerom) [reply sensor/common/deduper/deduper.go:51]: Done in #8078

File: sensor/common/deduper/deduper.go (hunk 7 above)
Comment (lvalerom) [reply sensor/common/deduper/deduper.go:64]: Done in #8078

File: sensor/common/sensor/central_communication_test.go [hunk 10]
```diff
@@ -142,6 +136,105 @@ func (c *centralCommunicationSuite) Test_StopCentralCommunication() {
...
+			go c.comm.(*centralCommunicationImpl).sendEvents(c.mockService, &reachable, c.mockHandler, c.mockDetector, c.comm.(*centralCommunicationImpl).receiver.Stop, c.comm.(*centralCommunicationImpl).sender.Stop)
+			c.mockService.connected.Wait()
+
+			for _, msg := range tc.componentMessages {
+				responsesC <- message.New(msg)
+			}
+
+			if tc.eventuallyMatchesState != nil {
+				c.Assert().Eventually(func() bool {
+					mu.Lock()
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/sensor/central_communication_test.go:203]: I tried both suggestions and I was still getting data races. I think it is because reading the slice to pass it to the function is already a read outside of the lock. Maybe I missed something though, let's move the conversation to #8084.

File: sensor/common/sensor/central_communication_test.go (hunk 8 above)
Comment (lvalerom) [reply sensor/common/sensor/central_communication_test.go:206]: Done in #8084

**Comment by lvalerom** [comment]
Closed in favor of:

* https://github.com/stackrox/stackrox/pull/8078
* https://github.com/stackrox/stackrox/pull/8084
* https://github.com/stackrox/stackrox/pull/8091
